File: packages/silo-client/src/transport/query-string.ts
import type { TransportQueryValue } from "./transport-request.js";

/**
 * Builds the `?a=1&b=2` suffix for a request. Omits `undefined` and `null`
 * values, JSON-encodes an object value (the `filter` AST) before
 * URL-encoding it, and returns `""` when nothing is set — so a call site can
 * concatenate the result onto a path unconditionally.
 */
export class QueryString {
  static build(query: Record<string, TransportQueryValue> | undefined): string {
    if (!query) return "";

    const parts: string[] = [];
    for (const [key, value] of Object.entries(query)) {
      if (value === undefined || value === null) continue;
      const encoded = typeof value === "object" ? JSON.stringify(value) : String(value);
      parts.push(`${encodeURIComponent(key)}=${encodeURIComponent(encoded)}`);
    }

    return parts.length > 0 ? `?${parts.join("&")}` : "";
  }
}

File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/transport/QueryString.java
package in.org.quicko.silo.client.transport;

import com.fasterxml.jackson.databind.JsonNode;
import java.util.Map;

/**
 * Builds the {@code ?a=1&b=2} suffix for a request. Omits a null value, renders
 * a {@link JsonNode} (the filter AST) as its compact JSON before encoding it,
 * and answers {@code ""} when nothing is set — so a call site can concatenate
 * the result onto a path unconditionally.
 */
public final class QueryString {
  private QueryString() {}

  public static String build(Map<String, Object> query) {
    if (query == null || query.isEmpty()) return "";

    StringBuilder built = new StringBuilder();
    for (Map.Entry<String, Object> parameter : query.entrySet()) {
      Object value = parameter.getValue();
      if (value == null) continue;
      String encoded = value instanceof JsonNode node ? node.toString() : String.valueOf(value);
      built.append(built.isEmpty() ? '?' : '&')
          .append(ApiPath.segment(parameter.getKey()))
          .append('=')
          .append(ApiPath.segment(encoded));
    }
    return built.toString();
  }
}
